include/circt/Analysis/FIRRTLInstanceInfo.h

```diff
@@ -0,0 +1,115 @@
+//===- FIRRTLInstanceInfo.h - Instance info analysis ------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the InstanceInfo analysis.  This is an analysis that
+// depends on the InstanceGraph analysis, but provides additional information
+// about FIRRTL operations.  This is useful if you find yourself needing to
+// selectively iterate over parts of the design.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef CIRCT_DIALECT_FIRRTL_FIRRTLINSTANCEINFO_H
+#define CIRCT_DIALECT_FIRRTL_FIRRTLINSTANCEINFO_H
+
+#include "circt/Dialect/FIRRTL/FIRRTLOps.h"
+#include "circt/Support/LLVM.h"
+#include "mlir/Pass/AnalysisManager.h"
+
+namespace circt {
+namespace firrtl {
+
+using namespace detail;
+
+class InstanceInfo {
+
+public:
+  explicit InstanceInfo(Operation *op, mlir::AnalysisManager &am);
+
+  /// A lattice value recording whether or not a property is true for a set of
+  /// operations.
+  struct LatticeValue {
+    enum Kind {
+      // Indicates that we have no information about this instance.  If the
+      // circuit has been completely analyzed, then this is equivalent to the
+      // property holds for zero instances.
+      Unknown = 0,
+      // Indicates that the property has a true or false value for all instances
+      // observed.
+      Constant,
+      // Indicates that the property holds for some, but not all instances.
+      Mixed
+    };
+
+    /// Whether or not the property holds.
+    Kind kind = Unknown;
+
+    /// The value of the property if `kind` is `Constant`.
+    bool constant = false;
+
+    // Combine this lattice value with another lattice value.
+    void merge(LatticeValue that);
+
+    // Merge evidence that a property is true.
+    void merge(bool property);
+  };
+
+  struct ModuleAttributes {
+    /// Indicates that this module is the desgin-under-test.  This is indicated
+    /// with a `sifive.enterprise.firrtl.MarkDUTAnnotation`.
+    bool isDut;
+
+    /// Indicates if this module is instantiated under the design-under-test.
+    InstanceInfo::LatticeValue underDut = {LatticeValue::Kind::Unknown};
```

Comment:
I think this interpretation of `underDut` makes more sense.

Reply:
I switched to some very exact language in 09b03f18a89cab75f3b553652b5ff85ed7b9d0a4. This language could hint the new interpretation of "under" isn't quite right. Specifically, is an instance of the DUT under the DUT? I went back and forth here and settled on it is under the DUT, i.e., the new interpretation makes sense.

Changed in 0d28c7aec.

WDYT?

Reply:
After further thought on this, I changed to use "any" instead of "atLeastOne" in fa3de1e7d2456d9f01cbd9023d00a0ba985709c3. This tends to better match, for lack of better inspiration, functional programming style "any of" or "all of" methods. "any" is also much shorter.